File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/MatchedDelayFunction.java
package cz.fb.manaus.reactor.profit.progress.function;

import cz.fb.manaus.core.model.SettledBet;
import org.springframework.stereotype.Component;

import java.time.temporal.ChronoUnit;
import java.util.OptionalDouble;

@Component
public class MatchedDelayFunction implements ProgressFunction {

    @Override
    public OptionalDouble apply(SettledBet bet) {
        var placed = bet.getPlacedOrActionDate().toInstant();
        if (bet.getMatched() != null) {
            var matched = bet.getMatched().toInstant();
            return OptionalDouble.of((double) placed.until(matched, ChronoUnit.MINUTES));
        }
        return OptionalDouble.empty();
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/MatchedDelayFunction.kt
package cz.fb.manaus.reactor.profit.progress.function

import cz.fb.manaus.core.model.RealizedBet
import org.springframework.stereotype.Component
import java.time.temporal.ChronoUnit

@Component
object MatchedDelayFunction : ProgressFunction {

    override fun invoke(bet: RealizedBet): Double? {
        val placed = bet.settledBet.placed
        if (placed != null) {
            val matched = bet.settledBet.matched
            return placed.until(matched, ChronoUnit.MINUTES).toDouble()
        }
        return null
    }
}
